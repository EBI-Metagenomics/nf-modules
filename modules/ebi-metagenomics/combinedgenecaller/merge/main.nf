process COMBINEDGENECALLER_MERGE {
    tag "${meta.id}"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/mgnify-pipelines-toolkit:1.0.1--pyhdfd78af_0':
        'biocontainers/mgnify-pipelines-toolkit:1.0.1--pyhdfd78af_0' }"

    input:
    tuple val(meta), path(pyrodigal_gff, stageAs: "pyrodigal/"), path(pyrodigal_ffn, stageAs: "pyrodigal/"), path(pyrodigal_faa, stageAs: "pyrodigal/"), path(fgs_gff, stageAs: "fgsrs/"), path(fgs_ffn, stageAs: "fgsrs/"), path(fgs_faa, stageAs: "fgsrs/"), path(mask)

    output:
    tuple val(meta), path("*.faa.gz"), emit: faa
    tuple val(meta), path("*.ffn.gz"), emit: ffn
    tuple val(meta), path("*.gff.gz"), emit: gff
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    // Check if files are compressed
    def is_prod_gff_compressed = pyrodigal_gff.name.endsWith(".gz")
    def is_prod_ffn_compressed = pyrodigal_ffn.name.endsWith(".gz")
    def is_prod_faa_compressed = pyrodigal_faa.name.endsWith(".gz")
    def is_fgs_gff_compressed = fgs_gff.name.endsWith(".gz")
    def is_fgs_ffn_compressed = fgs_ffn.name.endsWith(".gz")
    def is_fgs_faa_compressed = fgs_faa.name.endsWith(".gz")

    // Get uncompressed file names
    def pyrodigal_gff_file = pyrodigal_gff.name.replace(".gz", "")
    def pyrodigal_ffn_file = pyrodigal_ffn.name.replace(".gz", "")
    def pyrodigal_faa_file = pyrodigal_faa.name.replace(".gz", "")
    def fgs_gff_file = fgs_gff.name.replace(".gz", "")
    def fgs_ffn_file = fgs_ffn.name.replace(".gz", "")
    def fgs_faa_file = fgs_faa.name.replace(".gz", "")

    def is_mask_compressed = false
    def mask_parameter = ""
    if ( mask && mask.size() > 0 ) {
        is_mask_compressed = mask.name.endsWith(".gz")
        mask_parameter = "--mask ${mask.name.replace(".gz", "")} "
    }
    """
    if [ "${is_prod_gff_compressed}" == "true" ]; then
        gzip -d -c ${pyrodigal_gff} > ${pyrodigal_gff_file}
    fi
    if [ "${is_prod_ffn_compressed}" == "true" ]; then
        gzip -d -c ${pyrodigal_ffn} > ${pyrodigal_ffn_file}
    fi
    if [ "${is_prod_faa_compressed}" == "true" ]; then
        gzip -d -c ${pyrodigal_faa} > ${pyrodigal_faa_file}
    fi
    if [ "${is_fgs_gff_compressed}" == "true" ]; then
        gzip -d -c ${fgs_gff} > ${fgs_gff_file}
    fi
    if [ "${is_fgs_ffn_compressed}" == "true" ]; then
        gzip -d -c ${fgs_ffn} > ${fgs_ffn_file}
    fi
    if [ "${is_fgs_faa_compressed}" == "true" ]; then
        gzip -d -c ${fgs_faa} > ${fgs_faa_file}
    fi
    if [[ "${is_mask_compressed}" == "true" ]]; then
        gunzip ${mask}
    fi

    combined_gene_caller_merge \\
        ${args} \\
        -n ${prefix} \\
        --pyrodigal-gff ${pyrodigal_gff_file} \\
        --pyrodigal-ffn ${pyrodigal_ffn_file} \\
        --pyrodigal-faa ${pyrodigal_faa_file} \\
        --fgsrs-gff ${fgs_gff_file} \\
        --fgsrs-ffn ${fgs_ffn_file} \\
        --fgsrs-faa ${fgs_faa_file} ${mask_parameter}

    gzip -n ${prefix}.{faa,ffn,gff}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
         mgnify-pipelines-toolkit: \$(get_mpt_version)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.{faa,ffn,gff}

    gzip -n ${prefix}.{faa,ffn,gff}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
         mgnify-pipelines-toolkit: \$(get_mpt_version)
    END_VERSIONS
    """
}
