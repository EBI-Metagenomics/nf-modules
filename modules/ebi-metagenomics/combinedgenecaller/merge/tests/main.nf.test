nextflow_process {

    name "Test Process COMBINEDGENECALLER_MERGE"
    script "../main.nf"
    process "COMBINEDGENECALLER_MERGE"

    tag "modules"
    tag "modules_ebimetagenomics"
    tag "combinedgenecaller"
    tag "combinedgenecaller/merge"

    test("combinedgenecaller merge without a maskfile") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ],
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_pyrodigal.gff', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_pyrodigal.ffn', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_pyrodigal.faa', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_fgsrs.gff', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_fgsrs.ffn', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_fgsrs.faa', checkIfExists: true),
                    []
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("combinedgenecaller merge with a maskfile") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_pyrodigal.gff', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_pyrodigal.ffn', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_pyrodigal.faa', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_fgsrs.gff', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_fgsrs.ffn', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/input_fgsrs.faa', checkIfExists: true),
                    file(params.mgnify_pipelines_toolkit_testdata + '/combined_gene_caller_merge/cmsearch.all.tblout.deoverlapped', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
