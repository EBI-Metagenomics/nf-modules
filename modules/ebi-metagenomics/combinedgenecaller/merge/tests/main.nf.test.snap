{
    "combinedgenecaller merge without a maskfile": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,85bd33338eebe4c7f6a49121fe1b839c"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,dbe4300ce254687e182d89b94759c5f4"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.gff.gz:md5,774a5966f0cef53d3b17d4b5f432128e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,5aace80836214ecd3bdc321f7df46387"
                ],
                "faa": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,85bd33338eebe4c7f6a49121fe1b839c"
                    ]
                ],
                "ffn": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,dbe4300ce254687e182d89b94759c5f4"
                    ]
                ],
                "gff": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.gff.gz:md5,774a5966f0cef53d3b17d4b5f432128e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5aace80836214ecd3bdc321f7df46387"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-18T14:36:15.244102"
    },
    "combinedgenecaller merge with a maskfile": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,85bd33338eebe4c7f6a49121fe1b839c"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,dbe4300ce254687e182d89b94759c5f4"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.gff.gz:md5,774a5966f0cef53d3b17d4b5f432128e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,5aace80836214ecd3bdc321f7df46387"
                ],
                "faa": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,85bd33338eebe4c7f6a49121fe1b839c"
                    ]
                ],
                "ffn": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,dbe4300ce254687e182d89b94759c5f4"
                    ]
                ],
                "gff": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.gff.gz:md5,774a5966f0cef53d3b17d4b5f432128e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5aace80836214ecd3bdc321f7df46387"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-18T14:37:44.902928"
    }
}